cni/pkg/ambient/net.go

```diff
@@ -197,6 +202,77 @@ func buildRouteFromPod(pod *corev1.Pod, ip string) ([]string, error) {
 	}, nil
 }
 
+func buildEbpfArgsByIP(ip string, isZtunnel, isRemove bool) (*ebpf.RedirectArgs, error) {
+	ipAddr, err := netip.ParseAddr(ip)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse ip(%s): %v", ip, err)
+	}
+	veth, err := getVethWithDestinationOf(ip)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get device: %v", err)
+	}
+	peerIndex, err := getPeerIndex(veth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get veth peerIndex: %v", err)
+	}
+
+	peerNs, err := getNsNameFromNsID(veth.Attrs().NetNsID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get ns name: %v", err)
+	}
+
+	mac, err := getMacFromNsIdx(peerNs, peerIndex)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ebpf.RedirectArgs{
+		IPAddrs:   []netip.Addr{ipAddr},
+		MacAddr:   mac,
+		Ifindex:   veth.Attrs().Index,
+		PeerIndex: peerIndex,
+		PeerNs:    peerNs,
+		IsZtunnel: isZtunnel,
+		Remove:    isRemove,
+	}, nil
+}
+
+// GetHostIPByRoute get the automatically chosen host ip to the Pod's CIDR
+func GetHostIPByRoute(kubeClient kubernetes.Interface) (string, error) {
+	// We assume per node POD's CIDR is the same block, so the route to the POD
+	// from host should be "same". Otherwise, there may multiple host IPs will be
+	// used as source to dial to PODs.
+	pods, err := kubeClient.CoreV1().Pods(metav1.NamespaceAll).List(
+		context.TODO(),
+		metav1.ListOptions{
+			LabelSelector: "app=ztunnel",
+			FieldSelector: "spec.nodeName=" + NodeName,
+		})
+	if err != nil {
+		return "", fmt.Errorf("error getting ztunnel node: %v", err)
+	}
+	for _, pod := range pods.Items {
+		targetIP := pod.Status.PodIP
+		if hostIP := getOutboundIP(targetIP); hostIP != nil {
```

Comment:
Making all of these calls seems very expensive and suspicious. This is also changing a global variable which doesn't seem appropriate.

Why does

```
        - name: HOST_IP
          valueFrom:
            fieldRef:
              fieldPath: status.hostIP
```

not work?

Reply:
Directly use node's HOST_IP is not suitable for all scenarios, at least not work for kindnet or calico with force enabled ipip/vxlan mode. Thus, we emulate the behavior of kubelet on the host to get the proper "hostIP" which is used for health probe.

Reply:
Is there really no way to statically determine this without literally opening a connection?

Reply:
Ah TBH I thought this was querying the actual user pods rather than just ztunnel which makes it a bit better